packages/trie/src/verifyRangeProof.ts

```diff
@@ -0,0 +1,500 @@
+ *
+ * - One element proof. In this case no matter the edge proof is a non-existent
+ *   proof or not, we can always verify the correctness of the proof.
+ *
+ * - Zero element proof. In this case a single non-existent proof is enough to prove.
+ *   Besides, if there are still some other leaves available on the right side, then
+ *   an error will be returned.
+ *
+ * - Two edge elements proof. In this case two existent or non-existent proof(fisrt and last) should be provided.
+ *
+ * NOTE: Currently only supports verification when the length of firstKey and lastKey are the same.
+ *
+ * @param rootHash - root hash.
+ * @param firstKey - first key.
+ * @param lastKey - last key.
+ * @param keys - key list.
+ * @param values - value list, one-to-one correspondence with keys.
+ * @param proof - proof node list.
+ * @returns a flag to indicate whether there exists more trie node in the trie
+ */
+export async function verifyRangeProof(
+  rootHash: Buffer,
+  firstKey: Nibbles | null,
+  lastKey: Nibbles | null,
+  keys: Nibbles[],
+  values: Buffer[],
+  proof: Buffer[] | null
+) {
+  if (keys.length !== values.length) {
+    throw new Error('invalid keys and values')
+  }
+
+  // Make sure the keys are in order
+  for (let i = 0; i < keys.length - 1; i++) {
+    if (nibblesCompare(keys[i], keys[i + 1]) >= 0) {
+      throw new Error('invalid keys')
+    }
+  }
+  // Make sure all values are present
+  for (const value of values) {
+    if (value.length === 0) {
+      throw new Error('invalid values')
+    }
+  }
+
+  // All elements proof
+  if (proof === null) {
+    const trie = new Trie()
+    for (let i = 0; i < keys.length; i++) {
+      await trie.put(nibblesToBuffer(keys[i]), values[i])
+    }
+    if (rootHash.compare(trie.root) !== 0) {
+      throw new Error('invalid proof')
+    }
+    return false
+  }
+
+  if (firstKey === null || lastKey === null) {
+    throw new Error('invalid first or last key')
+  }
+
+  // Zero element proof
+  if (keys.length === 0) {
+    const { trie, value } = await verifyProof(rootHash, nibblesToBuffer(firstKey), proof)
+
+    if (value !== null || (await hasRightElement(trie, firstKey))) {
+      throw new Error('invalid proof')
+    }
+
+    return false
+  }
+
+  // One element proof
+  if (keys.length === 1 && nibblesCompare(firstKey, lastKey) === 0) {
+    const { trie, value } = await verifyProof(rootHash, nibblesToBuffer(firstKey), proof)
+
+    if (nibblesCompare(firstKey, keys[0]) !== 0) {
+      throw new Error('invalid proof')
+    }
+    if (value === null || value.compare(values[0]) !== 0) {
+      throw new Error('invalid proof')
+    }
+
+    return hasRightElement(trie, firstKey)
+  }
+
+  // Two edge elements proof
+
+  if (nibblesCompare(firstKey, lastKey) >= 0) {
+    throw new Error('invalid proof')
+  }
+  if (firstKey.length !== lastKey.length) {
+    throw new Error('invalid proof')
+  }
+
+  let trie = new Trie(null, rootHash)
+  trie = await Trie.fromProof(proof, trie)
```

Comment:
Here I notice geth is building the trie based on the `firstKey` and `lastKey`. Not sure if that makes a difference. If you've thought about it would appreciate a short explanation why there's no difference.

Reply:
The same position in geth is like this:

```go
root, _, err := proofToPath(rootHash, nil, firstKey, proof, true)
if err != nil {
  return false, err
}

root, _, err = proofToPath(rootHash, root, lastKey, proof, true)
if err != nil {
  return false, err
}
```

The logic of `proofToPath` is, resolve each node from `rootHash` through the input key, if the resolved node is `hashNode`, then link the child of `hashNode` to the parent of `hashNode`(in memory), until it reaches `valueNode` or the next child doesn't exist. Finally return the root node.

So geth does not build the trie based on `firstKey` and `lastKey`, but only resolves all necessary nodes based on `firstKey` and `lastKey`.

In js, we can use `trie.lookupNode` to resolve the next child node(it can be a raw node or a hash of the next child node). So instead of resolving all the nodes here, I use `trie.lookupNode` in the logic later to get the next child node.